language/move-prover/bytecode/src/loop_analysis.rs

```diff
@@ -54,43 +91,142 @@ impl FunctionTargetProcessor for LoopAnalysisProcessor {
+                                Bytecode::Prop(attr_id, PropKind::Assert, exp.clone())
                             });
                         }
+
+                        // havoc all loop targets
+                        for idx in &loop_info.val_targets {
+                            builder.emit_with(|attr_id| {
+                                Bytecode::Call(
+                                    attr_id,
+                                    vec![],
+                                    Operation::HavocVal,
+                                    vec![*idx],
+                                    None,
+                                )
+                            });
+                        }
+                        for (idx, havoc_all) in &loop_info.mut_targets {
+                            builder.emit_with(|attr_id| {
+                                Bytecode::Call(
+                                    attr_id,
+                                    vec![],
+                                    Operation::HavocRef(*havoc_all),
+                                    vec![*idx],
+                                    None,
+                                )
+                            });
+                        }
+
+                        // add an additional assumption that the loop did not abort
+                        let exp =
+                            builder.mk_not(builder.mk_bool_call(ast::Operation::AbortFlag, vec![]));
+                        builder.emit_with(|attr_id| Bytecode::Prop(attr_id, PropKind::Assume, exp));
+
+                        // re-assume loop invariants
+                        for (attr_id, exp) in loop_info.invariants.values() {
+                            builder.emit(Bytecode::Prop(*attr_id, PropKind::Assume, exp.clone()));
+                        }
                     }
                 }
+                Bytecode::Prop(_, PropKind::Assert, _)
+                    if invariant_locs.contains(&(offset as CodeOffset)) =>
+                {
+                    // skip it, as the invariant should have been added as an assert after the label
+                }
                 _ => {
                     builder.emit(bytecode);
                 }
             }
+            // mark that the goto labels in this bytecode needs to be updated to a new label
+            // representing the invariant-checking block for the loop.
+            if back_edge_locs.contains(&(offset as CodeOffset)) {
+                goto_fixes.push(builder.data.code.len() - 1);
+            }
+        }
+
+        // create one invariant-checking block for each fat loop
+        let invariant_checker_labels: BTreeMap<_, _> = loop_annotation
+            .fat_loops
+            .keys()
+            .map(|label| (*label, builder.new_label()))
+            .collect();
+
+        for (label, loop_info) in &loop_annotation.fat_loops {
+            let checker_label = invariant_checker_labels.get(label).unwrap();
```

Comment:
My personal preference is to keep them in the same function as the code represents the whole loop-to-DAG transformation algorithm, and it might hurt readability if we force some blocks into a helper function.

This `transform` is a bit different from the other large function `build_loop_annotation`. The latter can indeed be sliced into three helper functions, but it is hard to do so for `transform`.